### ISSUE TITLE
[Feature Request] EH Favorites Sync: Improved feedback during sync

### ISSUE THREAD

**Issue by lucky777709** [jobobby04/TachiyomiSY#356]
## Why/User Benefit/User Problem
For some months, I'm trying to sync my EH favorites. But the sync always "freezes" in a certain gallery.
With the current UI, I'm not sure what gallery the sync is trying to process, and I'm not sure if the app really froze (or if it may be just processing a gallery with lots of images). But the sync never seems to advance after that freezing.
Maybe some gallery is corrupted/invalid, but I'm not sure what gallery may be causing the freezing issue.

## What/Requirements
Some things that could add better feedback:
- An "Indeterminate Circular Progress Indicator" (Loading spinner), it would be constantly spinning to show that the sync is still running (and if the indicator suddenly froze, then I'd be sure the sync has been frozen).
- A "Determinate Linear Progress Indicator" (Loading bar), to help illustrate the overall progress of the sync.
- A dynamic text line displaying the name of the gallery currently being processed by the sync (similar to Windows installers). And if possible, also displaying what's exactly going on with the sync (maybe toggled by some "More Details" option).

**Comment by jobobby04** [comment]
I fixed the freeze issues in the newest preview build, so this isn't really relevant anymore 😄

**Comment by lucky777709** [comment]
"so this isn't really relevant anymore"
I disagree, I still think the sync would benefit from better feedback, at least displaying the name of the current gallery it's syncing.

I downloaded the latest preview version, but the sync still froze at a certain point, and now I don't know whether the sync crashed or not, and what's the problematic gallery.
In the message "Adding gallery X of Y to local library", the number of X now is greater than my previous tries (before it stopped at like "14 of 60-something", now it's at "43 of 64").
Maybe it froze at the exact same gallery as before (some problematic gallery that maybe I'd have to delete), but I can't be sure of that, based on the current UI.

**Comment by jobobby04** [comment]
Hmm, its probably a gallery with a crapton of previous versions, so techancally its not frozen, but is loading all previous versions as chapters. I guess its a good idea to display those if they take a while

**Comment by lucky777709** [comment]
Oh yes, that makes sense. I'd appreciate any kind of better feedback during the sync. :)

**Comment by lucky777709** [comment]
I recently upgraded to Preview r385, which displays the gallery name after a few seconds during the sync. This really improved the sync feedback. Thanks!

Thanks to this fix and other sync fixes, I managed to sync my favorites correctly.
@jobobby04 I guess you can close this if you wish.

**Comment by jobobby04** [comment]
Nice! I'll close it after the next release

**Comment by Accelyte** [comment]
Is it possible to add error messages when a gallery doesn't sync? The current feedback shows the name of the gallery, but you can't tell why it's failing on that particular gallery.

**Comment by jobobby04** [comment]
Not really no, it doesnt know its failing in the first place

**Comment by Accelyte** [comment]
Can you elaborate on what you mean by it doesn't know? Surely if a gallery is stuck, there should be reasons as to why it's stuck (e.g. gallery no longer exists, internet connection failure, etc.).

**Comment by jobobby04** [comment]
I guess that's technically wrong, everything other then many previous versions pause is properly handled by exceptions, which makes it skip that gallery. The only thing that *should* make it pause is that there are many previous versions, which actually will only take as long as it needs to get **every** version. There has been other issues in the past where it will pause the whole thread waiting on the db, meaning no work can be done, but that should be fixed.

**Comment by Accelyte** [comment]
Just as a suggestion (not sure if I should open a new issue): what if syncing only gets the latest version, and other previous versions are only retrieved when the gallery is viewed? That could speed up syncing large galleries (1000+).

**Comment by jobobby04** [comment]
That would require a few big changes in the code, and it's not a change that's worth it in the end in my opinion. Your best bet is the setting in the debug menu, which may or may not work.

**Comment by lucky777709** [comment]
I think there's still some underlying issue with the favorites sync, that makes that "long pause" take longer than it should (most times becoming stuck on certain galleries). I know a possible workaround for that.

In my case, rarely, there are still galleries that "freeze" during the favorites sync, and never finish syncing there. But, if I restart the app, go to that specific gallery (thanks again for displaying the gallery name during sync!), and refresh it manually (pull-to-refresh drag from the gallery title), it refreshes/updates the chapters VERY fast, and then the favorites sync works correctly with that gallery.

Maybe there's some loading process that becomes stuck during the favorites sync, but not when/after you manually refresh the gallery? This may help finding the real culprit of the sync freezing issue.

**Comment by Accelyte** [comment]
In my case, I had just switched from EhViewer, and I was trying to get all 1000+ favourites synced locally, so I can't even use your workaround. The app would just freeze, and I would have to restart the app completely, and even then nothing is synced. EhViewer didn't seem to have any issues with all these favourites, so I wonder if there's something wrong with the sync logic.
